sound/soc/sof/ipc4-pcm.c

```diff
@@ -249,9 +311,24 @@ static int sof_ipc4_pcm_dai_link_fixup(struct snd_soc_pcm_runtime *rtd,
 		return -EINVAL;
 	}
 
-	/* always set BE format to 32-bits for both playback and capture */
-	snd_mask_none(fmt);
-	snd_mask_set_format(fmt, SNDRV_PCM_FORMAT_S32_LE);
+	for_each_pcm_streams(dir) {
+		w = snd_soc_dai_get_widget(cpu_dai, dir);
+		if (w) {
+			struct snd_sof_widget *swidget = w->dobj.private;
+			struct snd_sof_widget *pipe_widget = swidget->pipe_widget;
+			struct sof_ipc4_pipeline *pipeline = pipe_widget->private;
+
+			if (pipeline->use_chain_dma)
+				use_chain_dma = true;
+		}
+	}
+
+	/* Always set BE format to 32-bits for both playback and capture, except if chained
+	 * DMA is in use. In chained DMA case the back-end and front-end format must match. */
```

Comment:
which begs the question: how is this match done and how are the codec_dai restrictions taken into account?

This is a similar problem to the dspless mode added by @ujfalusi

Reply:
@plbossart could you please spell out the codec DAI restriction we need to handle with chained DMA?

Reply:
@ranj063 @jsarha  see https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/sof/intel/hda-dai.c#L240

````c
	p_params.ch = params_channels(params);
	p_params.s_freq = params_rate(params);
	p_params.link_index = hlink->index;
	p_params.format = params_format(params);

	if (substream->stream == SNDRV_PCM_STREAM_PLAYBACK)
		p_params.link_bps = codec_dai->driver->playback.sig_bits;
	else
		p_params.link_bps = codec_dai->driver->capture.sig_bits;
````		

if the firmware does not modify the format, then don't wee need to change the format on the host side so that it matches what the codec can do?

Reply:
Ahh so we need to something similar to what @ujfalusi did as well ie save the format_val in hext_stream right?

Reply:
possibly, I am not sure what the firmware does and if this is a problem that needs solving, but the parallel with @ujfalusi 's work is interesting.

Reply:
I need a little digesting for this. ATM according to my tests everything else works OK with chained dma hdmi cpm, but only 48kHz is supported, and the PCM claims to support S24_LE format, but it does not.  The S24_LE samples are severely attenuated, probably due to the S24 being interpreted as S32 and 8msb bits being zero. Anyway I push the branch as it is now without addressing this at least to see if its otherwise good now.

Reply:
@jsarha can we add a comment in the commit message saying that only 16/32 bits are supported with chained DMA for now?

Reply:
so we need to change it in topology that only 16bit/32bit are support. FW doesn't change format so 24bit is a problem.